### DISCUSSION TITLE
Log Viewer

### DISCUSSION THREAD

**Discussion by sngazm** [lllyasviel/Fooocus#693] [Show and tell]
Hi, I am building a log viewer for myself. It parses original log.html and creates smart viewer for images.

![screen](https://github.com/lllyasviel/Fooocus/assets/4480229/22d2166b-6b88-4ca7-9c7c-0e9839efb72a)

Features:

- One HTML file. no additional backend, no npm install (it loads React and others from CDN)
- Quick date switching
- Auto reload new image (if today)
- Group images by "Batch" of image generation
- Show Diff with prompts from previous batch
- Sort images (newest first, oldest first)
- Mobile friendly
- Can choose how many images to load at once (important for mobile users like me)
- Easy grid resizing

If you place the attached viewer.html under the output directory, you can use the viewer by accessing 
 `/file=outputs/viewer.html` on the local server. Of course it doesn't send any data to outside.

[viewer.html.zip](https://github.com/lllyasviel/Fooocus/files/12909630/viewer.html.zip)

**Comment by toutjavascript** [comment]
Hi

I was looking for that kind of tool.
I suppose log format has changed since october. So, at line 486, 
`          const src = paragraphs[0].textContent;`
must be now
`	      const src = paragraphs[1].textContent;`

And it works great without error at start

Seed is not displayed yet. But easy to add.

I will work on it to add some functionnalities.

**Reply by xhoxye** [reply]
Very much looking forward to it

**Reply by toutjavascript** [reply]
Work in progress. 
I can copy seed to clipboard.
Next step is to integrate the log viewer in a new tab of fooocus. So it will be possible to drop all the datas (seed, prompt, models, ...) in the forms.

![Capture d’écran 2023-12-14 à 17 37 58](https://github.com/lllyasviel/Fooocus/assets/30899600/ce4b929c-febf-4ed2-8d45-3cc34060ca78)

**Reply by sngazm** [reply]
Thanks for using and fixing it! I forgot to upload it here, but here's the latest version of my own code I was using. Feel free to tweak it as you like.

[viewer_20231217.zip](https://github.com/lllyasviel/Fooocus/files/13692985/viewer_20231217.zip)
(update: merged @toutjavascript 's fix)

**Reply by bsdesigner** [reply]
I don't understand how it works or what folder to put it in. Maybe I have a problem with it because I'm working in Fooocus from Stability Matrix...?

**Reply by toutjavascript** [reply]
I checked Fooocus running in Stability Matrix. Viewer is running well.
Put the viewer.html file in \Downloads\StabilityMatrix-win-x64\Data\Packages\Fooocus\outputs
And then open http://localhost:7865/file=outputs/viewer.html

**Comment by toutjavascript** [comment]
I will publish a first release this week end
as I can see, Fooocus from Stability Matrix is more advance with viewer UI already build in

**Reply by toutjavascript** [reply]
So. Here some recent pictures from the viewer. It shows resolution and filesize on mouseover.
Click on the wen to view full screen pictures and datas

![Capture d’écran 2023-12-16 à 15 33 55](https://github.com/lllyasviel/Fooocus/assets/30899600/3757b869-d3d3-45bc-a201-555cba4a7ab2)

The datas are displayed on bottom. 
Click on each item to copy to clipboard. Paste every items in Fooocus UI and re-generate the very same image.

![Capture d’écran 2023-12-16 à 15 34 41](https://github.com/lllyasviel/Fooocus/assets/30899600/ad3d88c1-ca2e-4855-8433-1435e12cf3de)

I will publish here the viewer.html file in few hours.
Full Integration in Fooocus seams hard. If @lllyasviel wants, I could work with him to add viewer and made a one click button to get back all parameters.

**Comment by toutjavascript** [comment]
Here we go.

[viewer-20231216viewer.zip](https://github.com/lllyasviel/Fooocus/files/13693045/viewer-20231216viewer.zip)

Reminder : place the uncompressed viewer.html file in Fooocus/outputs directory (from run.bat)

You must open the viewer.html under Fooocus webserver (ie  http://localhost:7860/file=outputs/viewer.html or  http://localhost:7865/file=outputs/viewer.html )

I hope you will enjoy it.

**Comment by xhoxye** [comment]
Well done and thanks for contributing!
![QQ截图20231217053613](https://github.com/lllyasviel/Fooocus/assets/129571231/9adfd958-1d28-4dd8-8296-e0304ff9ad42)
![QQ截图20231217061602](https://github.com/lllyasviel/Fooocus/assets/129571231/d174ac12-0405-4564-8319-30dbcd3915ef)

**Reply by R4290N** [reply]
Hi xhoxye!
how to make this log viewer open from Focus UI?

**Comment by xhoxye** [comment]
A minor issue is that the zoomed-in display is not complete, cannot be dragged and dropped, and opens in another tab
![QQ截图20231217120100](https://github.com/lllyasviel/Fooocus/assets/129571231/37aaff7b-02ee-426c-81a3-1d3177579ee5)

**Reply by toutjavascript** [reply]
Thanx for testing.
You can reduce the metadata block by clicking on "Show/hide Metadatas"

I will optimize this display

**Reply by toutjavascript** [reply]
Keep working.
Image zoom size will be OK.
I will add direct iframe integration in Fooocus UI. This will make copy-paste of prompts very easy
It looks like :
![Capture d’écran 2023-12-17 à 15 17 00](https://github.com/lllyasviel/Fooocus/assets/30899600/e1db67e4-30eb-429c-9c29-03a32fabca87)

The main disadvantage is the install modification. We must add only one line in a .py file, but each time Fooocus is updated by @lllyasviel

**Reply by toutjavascript** [reply]
I was ready to publish the release. But the log format has changed. I must renew the html parser. And keep ancient one working to load previous log files.

**Reply by sngazm** [reply]
@toutjavascript 
Hi, I updated the `parseLog()` so that you can parse new and old logs.

[viewer_20231220.zip](https://github.com/lllyasviel/Fooocus/files/13717244/viewer_20231220.zip)

**Reply by toutjavascript** [reply]
Thanx @sngazm 

I created a new repository here to work on it
https://github.com/toutjavascript/Fooocus-Log-Viewer

First release is here, with deep iframe integration in Fooocus UI
https://github.com/toutjavascript/Fooocus-Log-Viewer/releases/tag/release

Do not hesitate to share ideas

**Reply by sngazm** [reply]
Great!

I'll refactor my code and start my repository too.

**Reply by toutjavascript** [reply]
hello @sngazm 
LoRA Log format is changed again. Can you modify the initial part you made (I can't find where you do that job)
Next version of Viewer Log with copy/paste metadatas is quite ready (with other functionnalities)

![Capture d’écran 2023-12-29 à 18 37 24](https://github.com/lllyasviel/Fooocus/assets/30899600/8312f5ae-3b3c-4f8c-81b6-2fc707a5ff00)

**Reply by sngazm** [reply]
@toutjavascript I'll look into the problem later. I'm busy during the year-end and New Year holidays, so I'm sorry if the resolution is delayed.

**Reply by sngazm** [reply]
I have submitted a pull request with the modification.
https://github.com/toutjavascript/Fooocus-Log-Viewer/pull/1

And here is my version.
https://github.com/sngazm/fooocus-viewer

**Reply by toutjavascript** [reply]
New release avalaible here
https://github.com/toutjavascript/Fooocus-Log-Viewer/releases/tag/release

Here some new features
![Capture d’écran 2023-12-31 à 11 56 09](https://github.com/lllyasviel/Fooocus/assets/30899600/02f13157-03f9-49ad-b510-8c2ee064a47a)

**Reply by toutjavascript** [reply]
I am still working on viewer.
I can fetch all ouputs directory to get only day with works.
For example in my Fooocus directory, I have that:
![Capture d’écran 2024-01-01 à 16 29 13](https://github.com/lllyasviel/Fooocus/assets/30899600/7776b37c-3ec6-4bcc-9b33-019306cb3d91)

The arrows on calendar jump at not empty days. But with React App, I don't understand how to pass the clicked date to input. Can you help @sngazm ?

**Reply by toutjavascript** [reply]
I am still working on the Log Viewer.
Here some new features :
At start a scan of all log directories is done to construct an array of day with images. The days and number of images can be displayed and clicks to directly go to the right day.
Arrows next/previous days are connected to the working day. No more navigation on an empty data day

![Capture d’écran 2024-01-07 à 15 35 24](https://github.com/lllyasviel/Fooocus/assets/30899600/69e1929f-804b-467f-b395-91ef4009b883)

Next release tomorrow

**Reply by toutjavascript** [reply]
Release is delivered here
https://github.com/toutjavascript/Fooocus-Log-Viewer/releases/tag/V1.2

Hope you will use it as often as me

**Reply by toutjavascript** [reply]
Still working with styles preview on log viewer:

![Capture d’écran 2024-01-09 à 16 45 01](https://github.com/lllyasviel/Fooocus/assets/30899600/540eb488-d13d-4ca0-aa3d-a5b4823e6228)

PS: same illustrations on models/loras would be great

**Reply by toutjavascript** [reply]
Working hard again	

![Capture d’écran 2024-01-11 à 19 33 44](https://github.com/lllyasviel/Fooocus/assets/30899600/22f75784-7ae4-4722-97d3-09c5c53b4a5e)

New release tomorrow i hope

**Reply by toutjavascript** [reply]
V1.3 is available
https://github.com/toutjavascript/Fooocus-Log-Viewer/releases/tag/v1.3

1.3.0 (thanks @ItsNoted for some of these ideas)
- Images of style are displayed on batch details
- Click on image thumbnail open zoomed image without targeting the magnifying glass
- Check for new image on today today's folder
- Could play sound when a new image is detected on today's folder
- Remember configuration choices (grid size, autoreload, sound on, display details, display metadata, ...)

**Reply by toutjavascript** [reply]
Hello

I am working again on log viewer with a Search Box to find images with text prompt, models and styles. 
There is no need to navigate by day to find your work
![Capture d’écran 2024-01-22 à 16 16 03](https://github.com/lllyasviel/Fooocus/assets/30899600/6bd7c198-61d5-410e-aa72-4a4bf7a34802)

Need some days to finish that work

**Reply by toutjavascript** [reply]
Here the V1.5 Release with search text into all prompts
https://github.com/toutjavascript/Fooocus-Log-Viewer/releases/tag/V1.5

Works very well, with over 10k images in outputs folder !
![Capture d’écran 2024-01-24 à 11 25 15](https://github.com/lllyasviel/Fooocus/assets/30899600/5861bb10-4876-43d6-a957-c935ab9a6929)

**Comment by R4290N** [comment]
> Главный недостаток – установка модификации. Мы должны добавить только одну строку в файл .py, но каждый раз, когда Fooocus обновляется

Please provide instructions

**Reply by xhoxye** [reply]
The higher the number of images, the slower each click opens the load

**Reply by Amit30swgoh** [reply]
Thanks for all the help bro you and Manuel Schmid are helping alot!
